### PR TITLE
Port TG Resonator so this is not a trash anymore

### PR DESCRIPTION
**Pull request by ErdGinalD**
## What Does This PR Do

Port resonator behaviour from TG https://github.com/tgstation/tgstation/pull/65297, now it can wave-spread after exposions up to 2 times, set a matrix traps who damage anyone who passes by and overall now its a good item for mining, like other mining tools.
- Base version now have 2 modes: manual and 2s exposion. Manual fields have 60s lifespan and not explode untill you break them manually. Other mode is self-explained, they explode after 2 seconds.
- Upgraded version now can set matrix traps, who immediately explode after someone will step by. Perfect for straight-line fights, like a dragon or vetus. Lifespan 1 minute.

## Why It's Good For The Game

Resonators are official trash no one wanted to deal with, adding a resonating wave and traps will make it reasonable choise between other tools like cutter or crusher. Still worse than explosives in term of mining ore, but they are too strong to compare with.

## Images of changes
From TG
https://gyazo.com/325c120bd1c7ee510634a7275f42731f

resonator wave
image
image
image

Matrix traps have different sprite. Once someone step on it, he will be fucked up.
image

## Testing

Run server, have an issue with matrix traps, fixed it, do few runs on my character setup to compare mining ability and megafauna killing potential, was satisfyed with words "i love this shit now".

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

:cl:
tweak: Resonator was reworked. Now it has a resonating wave after explosion, and upgraded version have ability to set traps. Once someone step on it, field will explode.
/:cl:

### CODE REVIEW

**Review by 1080pCat — APPROVED** [review]
i dont see any glaring issues.

**Comment by thraes** [comment]
Excited for this! :)

**Comment by Komrad822** [comment]
Need numbers, how much damage does the trap do to someone if it goes off, how much damage does it do to fauna.

File: code/modules/mining/equipment/resonator.dm [hunk 1]
```diff
@@ -9,42 +9,60 @@
...
+	if(mode == RESONATOR_MODE_AUTO)
+		to_chat(user, "<span class='notice'>You set the [name]'s fields to detonate only after you hit one with it.</span>")
+		mode = RESONATOR_MODE_MANUAL
+	else if(mode == RESONATOR_MODE_MANUAL)
+		to_chat(user, "<span class='notice'>You set the [name]'s fields to work as matrix traps.</span>")
+		mode = RESONATOR_MODE_MATRIX
 	else
-		burst_time = 50
-		to_chat(user, "<span class='notice'>You set the resonator's fields to detonate after 5 seconds.</span>")
+		to_chat(user, "<span class='notice'>You set the [name]'s fields to detonate after 2 seconds.</span>")
```
Comment (lewcc) [inline code/modules/mining/equipment/resonator.dm:48]:
```suggestion
		to_chat(user, "<span class='notice'>You set [src]'s fields to detonate automatically after 2 seconds.</span>")
```

File: code/modules/mining/equipment/resonator.dm [hunk 2]
```diff
@@ -53,31 +71,48 @@
...
-	transform = matrix() * 0.75
-	animate(src, transform = matrix() * 1.5, time = duration)
+	parent_resonator = set_resonator
+	if(parent_resonator)
+		parent_resonator.fields += src
+	adding_failure = set_failure
+	playsound(src, 'sound/weapons/resonator_blast.ogg', 50, TRUE)
+	if(mode == RESONATOR_MODE_AUTO)
+		transform = matrix()*0.75
+		animate(src, transform = matrix()*1.5, time = duration)
```
Comment (lewcc) [inline code/modules/mining/equipment/resonator.dm:106]:
```suggestion
		transform = matrix() * 0.75
		animate(src, transform = matrix() * 1.5, time = duration)
```

File: code/modules/mining/equipment/resonator.dm [hunk 3]
```diff
@@ -91,29 +126,42 @@
...
-	if(ismineralturf(T))
-		if(is_ancient_rock(T))
-			visible_message("<span class='notice'>This rock appears to be resistant to all mining tools except pickaxes!</span>")
-		else
-			var/turf/simulated/mineral/M = T
-			M.gets_drilled(creator)
-	check_pressure(T)
-	playsound(T,'sound/weapons/resonator_blast.ogg',50,1)
-	for(var/mob/living/L in T)
+	SIGNAL_HANDLER
```
Comment (lewcc) [inline code/modules/mining/equipment/resonator.dm:129]:
```suggestion
	SIGNAL_HANDLER  // COMSIG_MOVABLE_CROSSED
```

File: code/modules/mining/equipment/resonator.dm [hunk 4]
```diff
@@ -91,29 +126,42 @@
...
+		mineral_turf.gets_drilled(creator)
+	check_pressure(src_turf)
+	playsound(src_turf, 'sound/weapons/resonator_blast.ogg', 50, TRUE)
+	for(var/mob/living/attacked_living in src_turf)
 		if(creator)
-			add_attack_logs(creator, L, "Resonance field'ed")
-		to_chat(L, "<span class='userdanger'>[src] ruptured with you in it!</span>")
-		L.apply_damage(resonance_damage, BRUTE)
+			log_attack(creator, attacked_living, "used a resonator field on", "resonator")
+			SEND_SIGNAL(creator, COMSIG_LIVING_RESONATOR_BURST, creator, attacked_living)
```
Comment (lewcc) [inline code/modules/mining/equipment/resonator.dm:143]: If this signal has no listeners, we don't need it being sent. Please remove it for now.

Comment (ErdGinalD) [reply code/modules/mining/equipment/resonator.dm]: yeah, it use for style proc, but i actually didnt find one. Donezo

**Review by Burzah — APPROVED** [review]

File: code/modules/mining/equipment/resonator.dm [hunk 5]
```diff
@@ -9,42 +9,60 @@
...
 	var/fieldlimit = 4
+	/// the list of currently active fields from this resonator
 	var/list/fields = list()
-	var/quick_burst_mod = 0.8
+	/// the number that is added to the failure_prob, which is the probability of whether it will spread or not
+	var/adding_failure = 50
+
+/obj/item/resonator/attack_self(mob/user)
+	if(mode == RESONATOR_MODE_AUTO)
+		to_chat(user, "<span class='notice'>You set the [name]'s fields to detonate only after you hit one with it.</span>")
```
Comment (DGamerL) [inline code/modules/mining/equipment/resonator.dm:26]:
```suggestion
		to_chat(user, "<span class='notice'>You set the [name]'s fields to detonate only after you hit it with [src].</span>")
```

File: code/modules/mining/equipment/resonator.dm [hunk 6]
```diff
@@ -9,42 +9,60 @@
...
+		mode = RESONATOR_MODE_MANUAL
+	else
+		to_chat(user, "<span class='notice'>You set the [name]'s fields to detonate after 2 seconds.</span>")
```
Comment (DGamerL) [inline code/modules/mining/equipment/resonator.dm:29]:
```suggestion
		to_chat(user, "<span class='notice'>You set [src]'s fields to detonate after 2 seconds.</span>")
```

File: code/modules/mining/equipment/resonator.dm [hunk 7]
```diff
@@ -9,42 +9,60 @@
...
-/obj/item/resonator/attack_self(mob/user)
-	if(burst_time == 50)
-		burst_time = 30
-		to_chat(user, "<span class='notice'>You set the resonator's fields to detonate after 3 seconds.</span>")
+/obj/item/resonator/upgraded/attack_self(mob/user)
+	if(mode == RESONATOR_MODE_AUTO)
+		to_chat(user, "<span class='notice'>You set [src]'s fields to detonate only after being attacked by [src].</span>")
+		mode = RESONATOR_MODE_MANUAL
+	else if(mode == RESONATOR_MODE_MANUAL)
+		to_chat(user, "<span class='notice'>You set [src]'s fields to work as matrix traps.</span>")
```
Comment (DGamerL) [inline code/modules/mining/equipment/resonator.dm:45]: Maybe name it something like `proximity traps`? `matrix traps` sound very undescriptive

**Review by DGamerL — CHANGES_REQUESTED** [review]

File: code/modules/mining/equipment/resonator.dm (hunk 7 above)
Comment (ErdGinalD) [reply code/modules/mining/equipment/resonator.dm:45]: image
this already mention on upgraded resonator description, not sure if it need to change into something else.

**Review by DGamerL — APPROVED** [review]

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm [hunk 8]
```diff
@@ -537,7 +537,7 @@
 			R.damage_multiplier = modifier
 			R.burst()
 			return
-		new /obj/effect/temp_visual/resonance(target_turf, K.firer, null, 30)
+		new /obj/effect/temp_visual/resonance(target_turf, K.firer, null, RESONATOR_MODE_MANUAL, 100) //manual detonate mode and will NOT spread
```
Comment (1080pCat) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm:540]:
```suggestion
		new /obj/effect/temp_visual/resonance(target_turf, K.firer, null, RESONATOR_MODE_MANUAL, 100) // manual detonate mode and will NOT spread
```

**Pull request merged**